android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/KeyboardController.java

```diff
@@ -197,9 +197,12 @@ public boolean keyboardExists(String key) {
   }
 
   /**
-   * Add a new keyboard to the keyboard list. If the keyboard already exists, the keyboard
-   * information is updated.
-   * @param newKeyboard
+   * Add a new keyboard to the keyboard list, or update existing keyboard.
+   * If an existing keyboard is an exact match (language ID and keyboard ID), update the keyboard info.
+   * If the new keyboard package ID is not "cloud", then loop through the keyboard list for matching keyboard ID's:
```

Comment:
I'd prefer the cloud migration code to be in a separate module so that we can eliminate it easily later. That may mean refactoring but it should be done once on start of the app.

Reply:
Refactored to do migration in `KMManager.migrateCloudKeyboards()`